Let users download profiles as vCards

EvanHahn opened the pull request:
Users can now download a simple [vCard] for profiles they visit. This is implemented as a new route.

![App screenshot](https://github.com/Greening-Digital/constellate/assets/777712/47b3e01d-0ce5-4c39-96fd-86c6ba2d45ac)

Here's how it looks imported into [GNOME Contacts]:

![GNOME Contacts](https://github.com/Greening-Digital/constellate/assets/777712/a4e6aa2d-c410-41d8-a4bc-f5637488cfef)

And into iOS:

<p align="center"><img src="https://github.com/Greening-Digital/constellate/assets/777712/c932ca59-55fc-469a-a013-886b077db7fb"  alt="iOS screenshot" width="30%" /></p>

It uses [VObject], the best Python vCard package I could find. Some work was required to make edge cases work, such as the legacy vCard "name" field.

Addresses #106.

[vCard]: https://en.wikipedia.org/wiki/VCard
[VObject]: https://eventable.github.io/vobject/
[GNOME Contacts]: https://apps.gnome.org/Contacts/

EvanHahn committed:
Let users download profiles as vCards

Users can now download a simple [vCard] for profiles they visit. This is
implemented as a new route.

It uses [VObject], the best Python vCard package I could find. Some work
was required to make edge cases work, such as the legacy vCard "name"
field.

See [issue #106].

[vCard]: https://en.wikipedia.org/wiki/VCard
[VObject]: https://eventable.github.io/vobject/
[issue #106]: https://github.com/Greening-Digital/constellate/issues/106

EvanHahn commented on cl8/users/models.py:
```diff
@@ -284,6 +286,23 @@ def generate_invite_mail(self):
         return {"text": rendered_invite_txt, "html": rendered_invite_html}
 
+    def vcard(self) -> vobject.vCard:
+        properties = {
```
We could add additional properties in the future, but I think this is a good starter set. The [vCard Wikipedia page](https://en.wikipedia.org/wiki/VCard#Properties) has a good list of these properties, as does [the spec](https://datatracker.ietf.org/doc/html/rfc6350).

mrchrisadams commented on cl8/utils/tests/test_vcard_util.py:
```diff
@@ -0,0 +1,68 @@
+            "Biscuit",
+            "Dr.",
+            "Dr. Gravy Biscuit",
+            "Sweet Gravy Biscuit",
```
Good names.

mrchrisadams commented on cl8/utils/tests/test_vcard_util.py:
```diff
@@ -0,0 +1,68 @@
+
+
+class TestVcardUtil:
+    def test_full_name_to_name(self):
```
Would you mind adding a doc string giving a very short summary of what behaviour the test is checking is in place? Something like:

> check that we can free-form full name strings to structured Vcard names

or along those lines - assuming this is exercising `full_name_to_name()`

EvanHahn replied:
Done in 398261fabd55a5cb670fb9c88dd1ad6a43e20a06.

mrchrisadams commented on cl8/utils/tests/test_vcard_util.py:
```diff
@@ -0,0 +1,68 @@
+            expected = Name(given=given, family=family)
+            assert actual == expected
+
+    def test_content_disposition_filename(self):
```
Same here. I know the `content_disposition_filename()` is quite well documented, but a one liner along the lines of:

> check that we can create a valid filename for a vcard, based on a profile

Or something to that effect.

EvanHahn replied:
Done in 398261fabd55a5cb670fb9c88dd1ad6a43e20a06.

mrchrisadams reviewed (approved):
Thanks for this @EvanHahn , this looks great! I'm very happy to merge this in. - I've asked for a couple of docstrings for the tests as a convenience but otherwise I'm very happy for this to be merged in.

I'm not sure you gave the ability to merge in from an approved PR, if not please let me know, and I'll either grant access or merge it in at your end.

EvanHahn committed:
Add docstrings to vcard_util tests

EvanHahn commented:
I think I've addressed all comments!

I don't have permission to merge this myself. Happy to have you do it, or to get the permission to do so myself—whichever is easier for you.

mrchrisadams reviewed (commented):
Very happy! Thanks @EvanHahn
